### DISCUSSION TITLE
WalletConnect with telegram bot

### DISCUSSION THREAD

**Discussion by thanhnguyennguyen** [WalletConnect/walletconnect-monorepo#3507] [Developer Support]
I have an idea to build a telegram bot which pair with user wallet via walletConnect to make on chain transaction via telegram
My intended workflow is
- User  login their wallet using walletConnect (Telegram Bot will show an QR for them to scan and login)
- User type commands to make transaction
- TelegramBot makes an transaction and ask user Sign transaction via mobile wallet (supported WalletConnect)
- Broadcast transaction

However
AFAIK, WalletConnect is pairing with browser only 
is there any solution to pair a wallet ( supported WalletConnect) with Telegram Bot (basically it' a nodeJS backend)

**Comment by glitch-txs** [comment]
You can try this with sign API or core API from a backend

**Reply by thanhnguyennguyen** [reply]
Checking `core` package https://docs.walletconnect.com/2.0/api/core/pairing
But I'm still not clear how to generate QR code from backend for user to scan it and connect wallet

**Reply by glitch-txs** [reply]
You generate the QR code with the uri :)

**Reply by thanhnguyennguyen** [reply]
oh :) , thanks

**Reply by thanhnguyennguyen** [reply]
using typescript, install @walletconnect i got this error

```node_modules/@walletconnect/environment/dist/cjs/crypto.d.ts:1:44 - error TS2304: Cannot find name 'Crypto'.

1 export declare function getBrowerCrypto(): Crypto;
                                             ~~~~~~

node_modules/@walletconnect/environment/dist/cjs/crypto.d.ts:2:44 - error TS2304: Cannot find name 'SubtleCrypto'.

2 export declare function getSubtleCrypto(): SubtleCrypto;
                                             ~~~~~~~~~~~~


Found 2 errors in the same file, starting at: node_modules/@walletconnect/environment/dist/cjs/crypto.d.ts:1
```

any solution ?

**Reply by glitch-txs** [reply]
hmm @0xAsimetriq does core requires lokijs for backend? any workaround to this?

Maybe try adding some polyfill for SubtleCrypto interface, I'd assume is [this one](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto) 🤔

**Reply by xzilja** [reply]
I believe it needs it as alternative to localStorage yes.

**Reply by thanhnguyennguyen** [reply]
I'm installing this package in backend typescript and build failed
```
import { Core, Pairing, } from '@walletconnect/core'
```
so any solution to resolve above error

**Reply by rhlsthrm** [reply]
Don't use the core SDK, try something like this:

```ts
import AuthClient from "@walletconnect/auth-client";

const { topic, uri } = await authClient.core.pairing.create();
console.log("uri: ", uri);
```

I'm stuck on figuring out how to structure the URI into something that can be read by a wallet... I'm trying to also create a WalletConnect TG bot if you want to try to debug together @thanhnguyennguyen.

**Reply by thanhnguyennguyen** [reply]
yes, sure @rhlsthrm . I'm glad to hear that. let's debug together
I'm using typescript
Installed `"@walletconnect/auth-client"`
```bash  "dependencies": {
    "@influxdata/influxdb-client": "^1.33.2",
    "@walletconnect/auth-client": "^2.1.1",
    "axios": "^1.4.0",
    "config": "^3.3.9",
    "node-telegram-bot-api": "^0.61.0",
    "tslib": "~2.5",
    "winston": "^3.10.0"
  }
  ```
  
  Still got same error
  
  ```bash
  > tsc -p tsconfig.json

node_modules/@walletconnect/environment/dist/cjs/crypto.d.ts:1:44 - error TS2304: Cannot find name 'Crypto'.

1 export declare function getBrowerCrypto(): Crypto;
                                             ~~~~~~

node_modules/@walletconnect/environment/dist/cjs/crypto.d.ts:2:44 - error TS2304: Cannot find name 'SubtleCrypto'.

2 export declare function getSubtleCrypto(): SubtleCrypto;
                                             ~~~~~~~~~~~~


Found 2 errors in the same file, starting at: node_modules/@walletconnect/environment/dist/cjs/crypto.d.ts:1
```

**Reply by rhlsthrm** [reply]
I didn't have this error, maybe it's a node version thing?

**Reply by thanhnguyennguyen** [reply]
im using node v16.16.0
which node version you are using ?

**Comment by grzyslaw** [comment]
Hey, did you guys made any progress regarding the integration? If so, would you please be so kind to share some insights?

**Comment by Ydev1992** [comment]
Hi, so do you have telegram bot that do wallet connect just like gateway?

**Comment by christopheSeeka** [comment]
Can anyone who figured it share an example please? I would love to be able to do that too.

**Comment by andrew-cong** [comment]
a suck but work solution

https://docs.hotkeyswap.com/miscellaneous/connect-wallet-troubleshoot-guide/connecting-metamask-wallet-on-ios-a-step-by-step-guide

**Comment by mb-1208** [comment]
https://github.com/Reglament989/wc-webview-reproduction

Here's the handle for WalletConnect i guess.

**Comment by glitch-txs** [comment]
Hey everyone we are working on a nice solution for Telegram bots and MiniApps

**Reply by 0xpaperhead** [reply]
I faced a similar issue, I noticed it has to do with the way Telegram handles deep links and opening urls. ConnectKit seems to bypass this issue when it comes to wallet connections. Although triggering transactions could happening within the tApp, you have to go to the wallet manually go into the wallet and approve the transaction as it does not open the wallet automatically

**Reply by boneylva** [reply]
Awesome, is there any new news?

**Reply by alanwu4321** [reply]
@glitch-txs Are we waiting for this PR to merge? Is there anything we can do in the meantime to help or work around? 😄 

https://github.com/WalletConnect/web3modal/pull/2352
